libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -173,21 +173,74 @@ GCS_MAVLINK::setup_uart(const AP_SerialManager& serial_manager, AP_SerialManager
  * @brief Send the next pending waypoint, called from deferred message
  * handling code
  */
+void MissionItemProtocol::queued_request_send()
+{
+    if (link == nullptr) {
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+        AP_HAL::panic("Should always have a GCS_MAVLINK");
+#endif
+        return;
+    }
+    if (!receiving) {
+        return;
+    }
+    if (request_i > request_last) {
+        return;
+    }
+    mavlink_msg_mission_request_send(
+        link->get_chan(),
+        dest_sysid,
+        dest_compid,
+        request_i,
+        mission_type());
+    timelast_request = AP_HAL::millis();
+}
+
+void MissionItemProtocol::update()
+{
+    if (link == nullptr) {
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+        AP_HAL::panic("Should always have a GCS_MAVLINK");
+#endif
+        return;
+    }
+    const uint32_t wp_recv_time = 1000U + (link->get_stream_slowdown()*20);
+
+    // stop waypoint receiving if timeout
+    const uint32_t tnow = AP_HAL::millis();
+    if (tnow - timelast_receive > wp_recv_time+receive_timeout) {
```

Comment:
This can have time wrap problems. You need to not be adding the `wp_recv_time` limit here, it should be all managed via subtraction.

Reply:
Changed it to subtract receive_timeout on the LHS...